File: backend/src/main/java/com/gymmind/entity/Challenge.java
package com.gymmind.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "challenges")
public class Challenge {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long creatorId;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private String goalType; // WORKOUT_COUNT, TOTAL_TIME, TOTAL_CALORIES, DISTANCE

    @Column(nullable = false)
    private Integer goalValue;

    @Column(nullable = false)
    private LocalDate startDate;

    @Column(nullable = false)
    private LocalDate endDate;

    @Column(nullable = false)
    private String status; // UPCOMING, ACTIVE, COMPLETED

    @Column(nullable = false)
    private Integer participantCount;

    @Column(length = 500)
    private String imageUrl;

    @Column(length = 50)
    private String difficulty; // EASY, MEDIUM, HARD

    @CreationTimestamp
    private LocalDateTime createdAt;
}

File: frontend/src/api/challenge.ts
import request from '@/utils/request'

export interface CreateChallengeRequest {
  name: string
  description: string
  goalType: 'WORKOUT_COUNT' | 'TOTAL_TIME' | 'TOTAL_CALORIES' | 'DISTANCE'
  goalValue: number
  startDate: string
  endDate: string
  imageUrl?: string
  difficulty?: 'EASY' | 'MEDIUM' | 'HARD'
}

export interface ChallengeResponse {
  id: number
  creatorId: number
  name: string
  description: string
  goalType: string
  goalValue: number
  startDate: string
  endDate: string
  status: string
  participantCount: number
  imageUrl?: string
  difficulty?: string
  isParticipating: boolean
  userProgress?: number
  createdAt: string
}

export interface ParticipantResponse {
  userId: number
  username: string
  progress: number
  status: string
  rank: number
  joinedAt: string
}

// 创建挑战
export const createChallenge = (data: CreateChallengeRequest) => {
  return request<ChallengeResponse>({
    url: '/api/challenges',
    method: 'post',
    data
  })
}

// 获取活跃挑战
export const getActiveChallenges = () => {
  return request<ChallengeResponse[]>({
    url: '/api/challenges/active',
    method: 'get'
  })
}

// 获取用户参与的挑战
export const getUserChallenges = () => {
  return request<ChallengeResponse[]>({
    url: '/api/challenges/user',
    method: 'get'
  })
}

// 加入挑战
export const joinChallenge = (challengeId: number) => {
  return request({
    url: `/api/challenges/${challengeId}/join`,
    method: 'post'
  })
}

// 更新进度
export const updateProgress = (challengeId: number, progress: number) => {
  return request({
    url: `/api/challenges/${challengeId}/progress`,
    method: 'put',
    params: { progress }
  })
}

// 获取排行榜
export const getLeaderboard = (challengeId: number) => {
  return request<ParticipantResponse[]>({
    url: `/api/challenges/${challengeId}/leaderboard`,
    method: 'get'
  })
}
